cluster/saltbase/salt/kube-addons/initd

```diff
@@ -0,0 +1,123 @@
+#!/bin/bash
+#
+### BEGIN INIT INFO
+# Provides:   kube-addons
+# Required-Start:    $local_fs $network $syslog kube-apiserver
+# Required-Stop:
+# Default-Start:     2 3 4 5
+# Default-Stop:      0 1 6
+# Short-Description: Kubernetes Addon Object Manager
+# Description:
+#   Enforces installation of Kubernetes Addon Objects
+### END INIT INFO
+
+
+# PATH should only include /usr/* if it runs after the mountnfs.sh script
+PATH=/sbin:/usr/sbin:/bin:/usr/bin
+DESC="Kubernetes Addon Object Manager"
+NAME=kube-addons
+DAEMON_LOG_FILE=/var/log/$NAME.log
+PIDFILE=/var/run/$NAME.pid
+SCRIPTNAME=/etc/init.d/$NAME
+KUBECTL=/usr/local/bin/kubectl
+
+# Define LSB log_* functions.
+# Depend on lsb-base (>= 3.2-14) to ensure that this file is present
+# and status_of_proc is working.
+. /lib/lsb/init-functions
+
+function addon_manager_async() {
+    # The business logic for whether a given object should be created
+    # was already enforced by salt, and /etc/kubernetes/addons is the
+    # managed result is of that. Start everything below that directory.
+    echo "== Kubernetes addon manager started at $(date -Is) =="
+    for obj in $(find /etc/kubernetes/addons -name \*.yaml); do
+	    ${KUBECTL} create -f ${obj} &
+	    echo "++ addon ${obj} started in pid $! ++"
+    done
+    noerrors="true"
+    for pid in $(jobs -p); do
+	    wait ${pid} || noerrors="false"
+	    echo "++ pid ${pid} complete ++"
+    done
+    if [ ${noerrors} == "true" ]; then
+	    echo "== Kubernetes addon manager completed successfully at $(date -Is) =="
+    else
+	    echo "== Kubernetes addon manager completed with errors at $(date -Is) =="
+    fi
+
+    # We stay around so that status checks by salt make it look like
+    # the service is good. (We could do this is other ways, but this
+    # is simple.)
+    sleep infinity
```

Comment:
> # Required-Start: $local_fs $network $syslog kube-apiserver

Yes, it'll error if restarted for now. It's a bit of a wart.

What I really want is "create or update".


On Tue, Jan 20, 2015 at 10:43 AM, Joe Beda <notifications@github.com> wrote:

> In cluster/saltbase/salt/kube-addons/initd
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3601#discussion-diff-23244341>
> :
>
> > +    done
> > +    noerrors="true"
> > +    for pid in $(jobs -p); do
> > +	    wait ${pid} || noerrors="false"
> > +	    echo "++ pid ${pid} complete ++"
> > +    done
> > +    if [ ${noerrors} == "true" ]; then
> > +	    echo "== Kubernetes addon manager completed successfully at $(date -Is) =="
> > +    else
> > +	    echo "== Kubernetes addon manager completed with errors at $(date -Is) =="
> > +    fi
> > +
> > +    # We stay around so that status checks by salt make it look like
> > +    # the service is good. (We could do this is other ways, but this
> > +    # is simple.)
> > +    sleep infinity
>
> What happens if a user wants to push a new config for these things? I'm
> guessing that we don't support that for now. We might want to document that
> as part of kube-push.sh.
>
> Also, this'll error if the master machine reboots. We also need to make
> sure that on a reboot this guy comes up after the apiserver -- or perhaps
> waits/retries until the apiserver is ready.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3601/files#r23244341>
> .
>